src/lib/utils.ts

```diff
@@ -236,3 +231,114 @@ export function XOR(bool1: boolean, bool2: boolean): boolean {
 export function getActorNextEvents(snapshot: AnyMachineSnapshot) {
   return [...new Set([...snapshot._nodes.flatMap((sn) => sn.ownEvents)])]
 }
+
+export const onMouseDragRegex = /-?\.?\b\d+\.?\d*\b/g
+
+export function simulateOnMouseDragMatch(text: string) {
+  return text.match(onMouseDragRegex)
+}
+
+export function roundOff(num: number, precision: number = 2): number {
+  const x = Math.pow(10, precision)
+  return Math.round(num * x) / x
+}
+
+/**
+ * Determine if the number as a string has any precision in the decimal places
+ * '1' -> 0
+ * '1.0' -> 1
+ * '1.01' -> 2
+ */
+function getPrecision(text: string): number {
+  const wholeFractionSplit = text.split('.')
+  const precision =
+    wholeFractionSplit.length === 2 ? wholeFractionSplit[1].split('').length : 0
+  return precision
+}
+
+/**
+ * Determines if a number string has a leading digit
+ * 0.1 -> yes
+ * -0.1 -> yes
+ * .1 -> no
+ * 10.1 -> no
+ * The text.split('.') should evaluate to ['','<decimals>']
+ */
+export function hasLeadingZero(text: string): boolean {
+  const wholeFractionSplit = text.split('.')
+  return wholeFractionSplit.length === 2
+    ? wholeFractionSplit[0] === '0' || wholeFractionSplit[0] === '-0'
+    : false
+}
+
+function isWholeNumber(text: string): boolean | undefined {
+  const wholeFractionSplit = text.split('.')
+
+  if (wholeFractionSplit.length === 2) {
+    const wholeNumber = wholeFractionSplit[0]
+
+    if (wholeNumber.length === 0) {
+      return false
+    } else {
+      return true
+    }
+  }
+
+  if (wholeFractionSplit.length === 1) {
+    return true
+  }
+
+  // What if someone passes in 1..2.3.1...1.1.43
+  return undefined
+}
+
+export function onDragNumberCalculation(text: string, e: MouseEvent) {
+  const multiplier =
+    e.shiftKey && e.metaKey ? 0.01 : e.metaKey ? 0.1 : e.shiftKey ? 10 : 1
+
+  const delta = e.movementX * multiplier
+  const hasPeriod = text.includes('.')
+  const leadsWithZero = hasLeadingZero(text)
+  const addition = Number(text) + delta
+  const positiveAddition = e.movementX > 0
+  const negativeAddition = e.movementX < 0
+  const wholeNumber = isWholeNumber(text)
+  let precision = Math.max(
+    getPrecision(text),
+    getPrecision(multiplier.toString())
```

Comment:
I get what you're saying about string-to-number ping pong! But the logic is nicely clear here.